wip/beam-spec.md

```diff
@@ -0,0 +1,437 @@
+# The Basic Embedded Abstract Machine (BEAM)
+
+## Summary
+
+This document contains the specification of an *abstract machine*. The goal of
+this specification is facilitating discussions about the soundness of `unsafe`
+concurrent code that runs on microcontrollers.
+
+The BEAM is meant to be representative of the microcontrollers one can run Rust
+code on, as such it has concurrency capabilities commonly found in this kind of
+hardware but exposes these capabilities in a way that's not tied to any
+particular architecture.
+
+## Program execution
+
+- The BEAM has a single core / processing-unit and can only run a single process
+  on this core.
+
+- When the BEAM executes a program it uses the function called `main` as the
+  entry point. Returning from this function is UNSPECIFIED behavior.
+
+- The signature of `main` must be `[unsafe] fn() [-> !]`. That is the program
+  takes no arguments.
+
+## Interrupt handling
+
+(Interrupt handling is a preemption mechanism that in its most basic form works
+as follows: when an interrupt *signal* arrives the processor suspends the
+execution of the current function and jumps to a special subroutine known as the
+interrupt *handler*. When this subroutine returns the processors resumes the
+execution of the function that was preempted by the interrupt handler.)
+
+- BEAM's interrupt handling capabilities can be summarized as: prioritization of
+  interrupts, individual interrupt masking and global interrupt masking.
+
+- The BEAM has 8 *interrupts* whose *handlers* are named: `INTERRUPT0`,
+  `INTERRUPT1`, all the way to `INTERRUPT7`.
+
+- The signature of all interrupt handlers must be `[unsafe] fn()`. That is these
+  handlers don't take arguments nor can they return values.
+
+- If not defined in the program, interrupt handlers default to:
+
+``` rust
+#[no_mangle]
+unsafe fn INTERRUPT_() {
+    loop {}
+}
+```
+
+- When the BEAM starts, all interrupts are *disabled*, *unmasked* and are
+  configured with an initial *priority* of `1`; and the *running priority* of
+  the system is set to `0`.
```

Comment:
What is the difference between "disabled" and "masked"? I'd have thought these are the same thing...

It seems like you have some kind of per-interrupt state machine in mind here, with states like disabled, pending, masked, ...? If yes, then please write down that state machine explicitly.